File: controller.rb
class Controller 

	# construct
	def initialize(model, fields)
		@model = model
		@fields = fields
	end

	# creates a new controller
	# for codeiginiter framework
	def create
		file = File.new("ci/controllers/#{@model}.php", "w")
		file.write(content())
		file.close
		puts "\tController #{@model} was successfully created."
	end

	# generates the controller content
	# based on model name
	def content
  	"<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');\n\n" +
    "class #{@model.capitalize} extends CI_Controller {\n\n " +
    "\tfunction __construct() {\n " +
    "\t\tparent::__construct();\n " +
		"\t\t$this->load->model('#{@model.downcase}_model');\n" +
    "\t}\n\n" +
		"\tfunction index() {\n"+
		"\t\t$data['#{@model.downcase}'] = $this->#{@model.downcase}_model->findAll();\n" +
		"\t\t$this->load->view('#{@model.downcase}/index', $data);\n" +
		"\t}\n\n"+
		"\tfunction save() {\n" +
		get_fields_from_view +
		"\t\t$this->#{@model.downcase}_model->save($data)\n" +
		"\t\t$this->load->view('#{@model.downcase}/index', $data);\n" +
		"\t}\n" +
    "}\n " +
    "?>"
	end

	def get_fields_from_view
		sets = "\t\t$data[] = array();\n"
		@fields.each { |f|
			sets += "\t\tarray_push($this->input->xss_clean($this->input->post('#{f}')));\n"
		}
		sets.to_s
	end

end

File: lib/resources/controller.php
<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');

/**
 * <%=@model.capitalize%> Controller.
 */
class <%=@model.capitalize%> extends CI_Controller {

	function __construct() { 
		parent::__construct();
		$this->load->model('<%=@model.downcase%>_model');
	}

	# GET /<%=@model.downcase%>
	function index() {
		$data['<%=@model.downcase%>'] = $this-><%=@model.downcase%>_model->find();
		$data['content'] = '/<%=@model.downcase%>/index';
		$this->load->view('/includes/template', $data);
	}

	# GET /<%=@model.downcase%>/create
	function create() {
		$data['content'] = '/<%=@model.downcase%>/create';
		$this->load->view('/includes/template', $data);
	}

	# GET /<%=@model.downcase%>/edit/1
	function edit() {
		$id = $this->uri->segment(3);
		$data['<%=@model.downcase%>'] = $this-><%=@model.downcase%>_model->find($id);
		$data['content'] = '/<%=@model.downcase%>/create';
		$this->load->view('/includes/template', $data);
	}

	# GET /<%=@model.downcase%>/destroy/1
	function destroy() {
		$id = $this->uri->segment(3);
		$data['<%=@model.downcase%>'] = $this-><%=@model.downcase%>_model->destroy($id);
		redirect('/<%=@model.downcase%>/index', 'refresh');
	}

	# POST /<%=@model.downcase%>/save
	function save() {
		<% @attributes.each { |f| %>
		$this->form_validation->set_rules('<%= f.name %>', '<%= f.name.capitalize %>', 'required');<% } %>

		if ($this->form_validation->run()) {

			$data[] = array();
			$data['id'] = $this->input->post('id', TRUE);<% @attributes.each { |f| %>
			$data['<%= f.name %>'] = $this->input->post('<%= f.name %>', TRUE);<% } %>
			$this-><%=@model.downcase%>_model->save($data);
			redirect('/<%=@model.downcase%>/index', 'refresh');
		}
		$data['<%=@model.downcase%>'] =	$this->rebuild();
		$data['content'] = '/<%=@model.downcase%>/create';
		$this->load->view('/includes/template', $data);
	}

	function rebuild() {
		$object = new <%=@model.downcase%>_model();
		$object->id = $this->input->post('id', TRUE);<% @attributes.each { |f| %>
		$object-><%= f.name %> = $this->input->post('<%= f.name %>', TRUE);<% } %>
		return $object;
	}
}

?>

